src/libcmd/installable-flake.cc

```diff
@@ -28,6 +28,11 @@ namespace nix {
 std::vector<std::string> InstallableFlake::getActualAttrPaths()
 {
     std::vector<std::string> res;
+    if (attrPaths.size() == 1 && attrPaths.front().starts_with(".")){
```

Comment:
I've read the method docs, but I still don't really understand what this method does - crucially why it treats the first element specially.
We're going to need some documentation and/or refactoring here.

Reply:
Agreed, this is probably incorrect. Or at least needs justification.

Reply:
note: make this an assert

Reply:
This function seems to be called at times with the defaultPaths already present, namely when there is no attrPath component to the installable. (due to: https://github.com/NixOS/nix/blob/master/src/libcmd/installable-flake.cc#L76)

A possibility is to move the logic into the constructor. Something like:
```
    : InstallableValue(state),
      flakeRef(flakeRef),
      attrPaths(fragment == "" ? attrPaths : Strings{(std::string) fragment}),
      prefixes(fragment == "" || !fragment.starts_with(".") ? Strings{} : prefixes),
      extendedOutputsSpec(std::move(extendedOutputsSpec)),
      lockFlags(lockFlags)
```

Reply:
"An installable flake has attribute paths and prefixes" is a completely nonsensical statement to me, so I have no idea what the implications of this change would be without digging through everything.

Resolving installables is a procedure, not data, so I don't think attribute paths and prefixes should be in the same object in the first place.

So I guess this object is a representation of the internal state of that procedure, which is "fine", and what you're suggestion might work, And I guess what I'm trying to say is that a more functional or procedural style would help for cleaning up the existing code.